src/overlays/actors/ovl_En_Egol/z_en_egol.c

```diff
@@ -119,99 +316,1319 @@ const ActorInit En_Egol_InitVars = {
+                    Math_ApproachF(&this->chargeLightScale, 1.0f, 0.5f, 1.0f);
+                    if (this->waitTimer == 0) {
+                        this->chargeLevel = 0;
+                        this->laserState = EYEGORE_LASER_FIRE;
+                        this->chargingLaser = false;
+                        this->chargeLightScale = 0.0f;
+                    }
+                    break;
+            }
+            this->chargeLightRot += 0x7D0;
+        }
+        if ((this->laserState != EYEGORE_LASER_OFF) || (this->laserCount != 0)) {
+            Math_SmoothStepToS(&this->headRot, -0x2710, 5, 0x1F4, 5);
+        } else {
+            Math_SmoothStepToS(&this->headRot, 0, 5, 0x1F4, 5);
+        }
+        if (this->laserState == EYEGORE_LASER_OFF) {
+            if (this->laserCount >= 3) {
+                EnEgol_SetupWalk(this);
+            } else {
+                this->laserRot.x = sLaserAngles[this->laserCount];
+                this->laserScaleTarget.z = 0.0f;
+                this->laserScale.z = 0.0f;
+                if (this->laserCount == 0) {
+                    this->laserState = EYEGORE_LASER_START;
+                } else {
+                    this->laserState = EYEGORE_LASER_FIRE;
+                    this->laserScaleTarget.x = 0.03f;
+                    this->laserScaleTarget.y = 0.03f;
+                }
+                this->laserCount++;
+            }
+        } else if (this->laserState >= EYEGORE_LASER_FIRE) {
+            Actor_PlaySfxAtPos(&this->actor, NA_SE_EN_EYEGOLE_BEAM - SFX_FLAG);
+            if (this->laserState != EYEGORE_LASER_OFF) {
+                EnEgol_DestroyBlocks(this, play, this->laserCollider.dim.quad[0], this->laserCollider.dim.quad[1]);
+            }
+            if ((this->actionTimer == 0) && BgCheck_EntityLineTest1(&play->colCtx, &this->laserCollider.dim.quad[3],
+                                                                    &this->laserCollider.dim.quad[1], &hitPos, &colPoly,
+                                                                    false, true, false, true, &bgId)) {
+                //! @bug this should check walls, too
+                Vec3s rotToNorm;
+                f32 nx;
+                f32 ny;
+                f32 nz;
+                s32 pad3;
+                s32 i;
+                s32 quakeSize;
+                Player* player = GET_PLAYER(play);
+
+                nx = COLPOLY_GET_NORMAL(colPoly->normal.x);
+                ny = COLPOLY_GET_NORMAL(colPoly->normal.y);
+                nz = COLPOLY_GET_NORMAL(colPoly->normal.z);
+
+                rotToNorm.x = -func_80086B30(-nz * ny, 1.0f) * 0x8000 / M_PI;
+                rotToNorm.z = func_80086B30(-nx * ny, 1.0f) * 0x8000 / M_PI;
+                //! @bug This is the wrong calculation for the angles. It should be
+                // rotToNorm.x = func_80086B30(nz, ny) * 0x8000 / M_PI;
+                // rotToNorm.z = func_80086B30(-nx, sqrtf(1.0f - SQ(nx))) * 0x8000 / M_PI;
```

Comment:
Said it before, but in this case I'd prefer even *more* if this was above the code, not below it. Also, I'm not sure it's a bug per se? Lots of actors use this calculation

Reply:
Research in Discord suggests this is a bug (in that it causes unintended behaviour on steep slopes). I'd still like to know more about what it's supposed to do, though.